scripts/speech_recognition/oomptimizer.py

```diff
@@ -0,0 +1,525 @@
+                case "text", "audio":
+                    return (
+                        input_len,
+                        compute_num_samples(output_len, sampling_rate=sampling_rate),
+                    )
+                case "text", "text":
+                    return input_len, output_len
+                case _:
+                    raise RuntimeError(f"Unexpected modality combination: {_}")
+
+        return [_determine_lens_for_bucket(bin) for bin in buckets]
+
+    click.echo("Starting profiling.")
+    max_seq_lens = get_max_seq_lens(buckets)
+    gen = ProfilingBatchGenerator(schema=schema, start_batch_size=start_batch_size, rel_gap_thresh=threshold)
+    profile = {}
+
+    # Iterate buckets from the largest to the smallest sequences. This usually ends up creating
+    # a tiny bit smaller batches, likely due to worse memory fragmentation.
+    for bucket, (seq_len_in, seq_len_out) in reversed(list(zip(buckets, max_seq_lens))):
+        click.echo(f"The current sequence lengths are: input={seq_len_in} output={seq_len_out}.")
+        gen.reset()
+        batch_idx = 0
+
+        def step():
+            click.echo(
+                f"\t[BEGIN step] [CUDA RAM CURRENT: {torch.cuda.memory_allocated() / (1024 * 1024):.1f}MB] [CUDA RAM MAX: {torch.cuda.max_memory_allocated() / (1024*1024):.1f}MB]"
+            )
+            batch = gen(seq_len_in, seq_len_out)
+            oom = False
+            try:
+                click.echo(f"\tCurrent gap: {gen.current_rel_gap}... ", nl=False)
+                optimizer.zero_grad()
+                out = model.training_step(batch, batch_idx)
+                out['loss'].sum().backward()
+                optimizer.step()
+            except torch.cuda.OutOfMemoryError as e:
+                click.secho(f"OOM!", fg="yellow")
+                oom = True
+            except RuntimeError as e:
+                if "cuFFT error: CUFFT_INTERNAL_ERROR" not in str(e):
+                    raise
+                click.secho(f"OOM!", fg="yellow")
+                oom = True
+            else:
+                click.secho(f"OK!", fg="green")
+            finally:
+                click.echo(
+                    f"\t[END step] [CUDA RAM CURRENT: {torch.cuda.memory_allocated() / (1024 * 1024):.1f}MB] [CUDA RAM MAX: {torch.cuda.max_memory_allocated() / (1024*1024):.1f}MB]"
+                )
+                del batch
+                # Note: We could call empty_cache() to free up some more memory on the GPU,
+                #       but we have found out empirically that this causes a mismatched condition
+                #       between OOMptimizer and the actual training. During training, there is some
+                #       degree of memory fragmentation and it's better to simulate that in OOMptimizer.
+                # torch.cuda.memory.empty_cache()
+                torch.cuda.reset_max_memory_allocated()
+            return oom
+
+        with torch.autocast("cuda", getattr(torch, dtype)):
```

Comment:
Small FYI: the autocast's "cache cast" will be destroyed upon exiting the context manager's context. This should not affect the correctness of your work (since the new allocations will be the same size as the old ones and the allocator does a "best fit" search), but it does impose a slowdown because the cache must be reconstructed.

Reply:
cool, moved the autocast to cover the whole profiling loop in that case